rfc/kubernetes-ui-enhancements/kubernetes-ui-enhancements.md

```diff
@@ -0,0 +1,260 @@
+
+![](storage-mock.png)
+*The Spinnaker UI with a proposed Storage tab*
+
+![](pvc-mock.png)
+*A PVC listing mockup grouped by account and namespace.*
+
+### Config Maps
+
+#### Current Functionality
+
+Similar to its handling of Persistent Volume Claims, Spinnaker supports deploying Config Maps but does not render them in the UI or surface information on Config Map consumers. Other Kubernetes tools often feature a dedicated UI category for reviewing a cluster’s Config Maps.
+
+#### Proposal
+
+Similar to the effort to expose PVCs under a dedicated Storage tab, the UI components can be extended to surface Config Maps. These Config Maps will live in an additional UI tab for Configuration. A Config Maps’ stored data, a YAML viewer, and the following Metadata will be displayed:
+
+- Name
+- Namespace
+- Labels
+- Annotations
+- Created timestamp
+- Cluster
+
+### Jobs
+
+#### Current Functionality
+
+Although easily accessible in other Kubernetes dashboards, the jobs object is only indirectly viewable in Spinnaker’s infrastructure view. Instead, Spinnaker offers two distinct UI experiences when deploying Kubernetes jobs.
+
+##### Run Job Stage
+
+When executing a pipeline with a Run Job stage, an execution UI is available. The sole Job specific component of this view is a link to the container output.
+
+##### Deploy Manifest Stage
+
+When a Job is deployed using the Deploy Manifest stage, a more robust UI is available. This UI offers details covering deployment events and artifacts.
+
+![](deploy-manifest-job.png)
+*The UI for deploying a job through the Deploy Manifest stage*
+
+Additional metadata and job actions can be found by clicking the “Details” link. This takes the user to a panel on the infrastructure page that provides metadata on:
+
+- The job image
+- Labels
+- Size
+- Health
+
+Job actions for editing, scaling, or deleting the job are also available.
+
+![](job-metadata-panel.png)
+
+*The Job metadata panel as seen in the infrastructure UI*
+
+#### Proposal
+
+##### Display Job metadata on the infrastructure view
+
+The Job metadata panel serves as an effective means of surfacing the status and attributes of deployed jobs. However, it cannot be accessed directly. The metadata panel is not visible to users landing on the infrastructure route. Users must follow a link from a pipeline execution or  manipulate the URL by hand to find a given job. If the job was not deployed using Spinnaker or the execution history is unavailable, no link will exist to view the Job UI. Feedback from Spinnaker users, reveals that there is an expectation that the Job UI panel would appear in the infrastructure view. 
+
+The infrastructure view can be updated to display a list of jobs deployed to each cluster. Selecting a job from this list would then display the job metadata panel. This change would increase the visibility of deployed jobs and make it easier for users to monitor job resources across multiple clusters within Spinnaker.
+
```

Comment:
Question: I'm assuming that this UI will apply to batch resources deployed both by the deploy(manifest) stage and the run job (manifest) stage, right?

Request: If we are going to do this, can we ensure that in the UI, related jobs are associated together?  For example, if you are using the run job manifest stage, it will annotate a hash to the run job name, so it's technically a different run job - can we group these together (either via annotation, or something else?).

Reply:
So to my understanding, there are two types of `batch` objects in the Kubernetes API: Jobs and CronJobs

Independent of the above, I think I'm primarily concerned about batch objects where there may be more than one instance of the job running at a given time, and I think there are two high-level use cases for this:
* A pool of worker pods churning through some pool of jobs
* A given job that may be called from different pipeline executions (same job, different parameters; for example, each 'job' creates and monitors a ticket, which is a use case I've seen a lot recently)

Key point here: Job specs are immutable, so the way the RJ stage works is that Spinnaker 'versions' the RJ with an appended string.

I think CronJobs can be left for a later discussion - I think an ideal goal would be something like the following:
* CronJobs create and manage Jobs, similar to the way Deployments create and manage ReplicaSets, so you could have a 'parent' CronJob object that has child Job objects, each of which would have child Pods
* I think this would require building a bit of an abstraction layer on top of the CronJob object in Spinnaker (haven't dug too much into this), so maybe this is a longer-term goal, just something to keep in mind.

For (again, immutable) Jobs created via the "Deploy (Manifest)" stage, I think the behavior is relatively straightforward - group all of the pods for a given job together the same way the standalone un-versioned ReplicaSet object is displayed (with the knowledge that in many instance, there will probably only be one pod running at a time).

For Jobs created by the "Run Job (Manifest)" stage, Spinnaker currently appends an alphanumeric string to the job `name` before deploying it.  If there were a way to track the un-appended job name in the spec (annotation of some sort - `moniker.spinnaker.io/job`?), then we could have all the pods for a given job instance in one level, and then all the Jobs for a given job grouped together at a higher level.

I'm not sure I'm doing a great job of explaining this, but hopefully this makes sense.

Reply:
Thank you for the detailed feedback!

+1, Displaying the Jobs and Pods that were triggered by a Cron Job would be helpful. This may be achievable by grouping on the ownerReferences metadata available on Jobs triggered by Cron Jobs. 

Yeah, I think Jobs deployed through the Deploy Manifest stage should group pods together in the UI. I plan on having parallel jobs render their associated pods.

I like the idea of grouping the unique Job objects created by the Run Job Stage and leveraging annotations seems like a viable way to do so. I’m not as familiar with the internals of how those annotations are managed so this may end up being a follow up to my initial set of changes.

I’ve gone ahead and updated the RFC to reflect all this.